src/server/transaction.cc

```diff
@@ -549,75 +536,63 @@ void Transaction::MultiData::AddLocks(IntentLock::Mode mode) {
 // BLPOP where a data must be read from multiple shards before performing another hop.
 OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
   DCHECK(!cb_);
-
   cb_ = std::move(cb);
 
-  // single hop -> concluding.
-  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);
+  DCHECK(multi_ || (coordinator_state_ & COORD_SCHED) == 0);   // Only multi schedule in advance.
+  coordinator_state_ |= (COORD_EXEC | COORD_EXEC_CONCLUDING);  // Single hop means we conclude.
 
-  if (!multi_) {  // for non-multi transactions we schedule exactly once.
-    DCHECK_EQ(0, coordinator_state_ & COORD_SCHED);
-  }
-
-  bool schedule_fast = (unique_shard_cnt_ == 1) && !IsGlobal() && !multi_;
-  bool run_eager = false;
+  bool was_ooo = false;
 
-  if (schedule_fast) {  // Single shard (local) optimization.
-    // We never resize shard_data because that would affect MULTI transaction correctness.
+  // If we run only on one shard and conclude, we can avoid scheduling at all
+  // and dispatch a single task to the its shard set.
+  if ((unique_shard_cnt_ == 1) && !IsGlobal() && !multi_) {
     DCHECK_EQ(1u, shard_data_.size());
 
+    // IsArmedInShard() first checks run_count_ before shard_data, so use release ordering.
     shard_data_[0].local_mask |= ARMED;
-
-    // memory_order_release because we do not want it to be reordered with shard_data writes
-    // above.
-    // IsArmedInShard() first checks run_count_ before accessing shard_data.
     run_count_.fetch_add(1, memory_order_release);
+
     time_now_ms_ = GetCurrentTimeMs();
 
-    // Please note that schedule_cb can not update any data on ScheduleSingleHop stack when
-    // run_fast is false.
-    // since ScheduleSingleHop can finish before ScheduleUniqueShard returns.
-    // The problematic flow is as follows: ScheduleUniqueShard schedules into TxQueue
-    // (hence run_fast is false), and then calls PollExecute that in turn runs
-    // the callback which calls DecreaseRunCnt.
-    // As a result WaitForShardCallbacks below is unblocked before schedule_cb returns.
-    // However, if run_fast is true, then we may mutate stack variables, but only
-    // before DecreaseRunCnt is called.
-    auto schedule_cb = [&] {
+    // NOTE: schedule_cb cannot update data on stack when run_fast is false.
+    // This is because this function can finish before the callback returns.
```

Comment:
`this function` - not clear which one you are referring to. Maybe just use `ScheduleSingleHop`